parsers/CA_NS.py

```diff
@@ -1,213 +1,182 @@
 
-        # in mix_data, the values are expressed as percentages,
-        # and have to be multiplied by load to find the actual MW value.
-        corresponding_load = [
-            load_period
-            for load_period in load_data
-            if load_period["datetime"] == mix["datetime"]
-        ]
-        if corresponding_load:
-            load = corresponding_load[0]["Base Load"]
+        if timestamp in loads:
+            load = loads[timestamp]
         else:
-            # if not found, assume 1244 MW, based on average yearly electricity available for use
-            # in 2014 and 2015 (Statistics Canada table Table 127-0008 for Nova Scotia)
+            # If a base load corresponding with this timestamp is not found,
+            # assume 1244 MW based on the average yearly electricity available
+            # for use in 2014 and 2015 (Statistics Canada table 127-0008 for
+            # Nova Scotia).
             load = 1244
             logger.warning(
-                f"unable to find load for {data_date}, assuming 1244 MW",
-                extra={"key": zone_key},
+                f"unable to find load for {timestamp}; assuming 1244 MW",
+                extra={"key": ZONE_KEY},
             )
 
-        electricity_mix = {
-            gen_type: percent_value * load
-            for gen_type, percent_value in percent_mix.items()
-        }
-
-        # validate again
-        valid = True
-        for gen_type, value in electricity_mix.items():
-            absolute_bound = valid_absolute.get(
-                gen_type
-            )  # imports are not in valid_absolute
-            if absolute_bound and value > absolute_bound:
-                valid = False
-                logger.warning(
-                    "discarding datapoint at {dt} due to {fuel} "
-                    "too high: {value} MW".format(
-                        dt=data_date, fuel=gen_type, value=value
-                    ),
-                    extra={"key": zone_key},
-                )
-        if not valid:
-            # continue the outer loop, not the inner
-            continue
-
-        production.append(
-            {
-                "zoneKey": zone_key,
-                "datetime": data_date,
-                "production": {
-                    key: value
-                    for key, value in electricity_mix.items()
-                    if key != "imports"
-                },
-                "source": "nspower.ca",
-            }
+        # In this source, imports are positive. In the expected result for
+        # CA-NB->CA-NS, "net" represents a flow from NB to NS, i.e., an import
+        # to NS, so the value can be used directly. Note that this API only
+        # specifies imports; when NS is exporting energy, the API returns 0.
+        exchanges.append(
+            datetime=timestamp,
+            netFlow=load * mix["Imports"] / 100,
+            source=SOURCE,
+            zoneKey=ZoneKey("CA-NB->CA-NS"),
```

Comment:
Not strictly related to this refactor, but just to mention. The assumption that all imports are from NB is now problematic because Nova Scotia's interconnection to Newfoundland is now live. But I've never seen live data on its operation, the closest thing was that it amounted to 14% of NS supply in 2023H1, per https://www.nspower.ca/cleanandgreen/clean-energy "Our Energy Stats" chart

Reply:
For reference related to your comment: the daily generation reports for Newfoundland *might* be usable as a source for the interconnector data, see
https://github.com/electricitymaps/electricitymaps-contrib/issues/2541#issuecomment-1879357645

Reply:
Hey, look: I found a page the NSPower website that has the information you need (Maritime Link interconnector) and more, updated roughly every 5 minutes, with more granular data than the current source: [it's called Daily Report, but under the heading "Current System Conditions" you'll find "10 minute averages" of e.g. load, wind, imports and exports through various interconnectors that are at most five minutes old](https://www.nspower.ca/oasis/system-reports-messages/daily-report).

Reply:
So I dug deeper into this, and it turns out we aren't actually using this exchange data since [we get the data on the NS-NB interconnection from New Brunswick Power](https://github.com/electricitymaps/electricitymaps-contrib/blob/master/config/exchanges/CA-NB_CA-NS.yaml), not from NSPower. So it might be best to just delete the exchanges related part.

Reply:
Hey @VIKTORVAV99, any thoughts on this? Should we simply delete the exchange parser if it's not used?

Reply:
To clarify, my idea was to delete the exchange parser now so we are more certain we don't accidentally fetch incorrect data, and then re-add an exchange parser based on the new source from #6317 in a separate pull request.

But doing both in this PR is also acceptable, if you've got the time for it.

Reply:
Oh, gotcha! It's not entirely clear to me what needs to be done to implement the new exchange parser (there are a lot of links to follow and I'm a bit lost), so I'll just delete it here and we can address it separately. Would those changes have any impact on the production parser?

Reply:
The exchange stuff shouldn't. The "[Current System Conditions](https://www.nspower.ca/oasis/system-reports-messages/daily-report)" also has wind generation, but as the page has data updated roughly every 5 minutes I think it'll be hard to integrate with the hourly data provided by the current source &ndash; and probably not that useful either, if the current source stays online.

As for implementing the exchange parser, the short summary is: from https://www.nspower.ca/oasis/system-reports-messages/daily-report, under "**Current System Conditions**", grab:
1. The timestamp, e.g. "Last Updated: 10-Jan-24 20:38:10";
2. "NS Export" for (most likely) the New Brunswick interconnector (optional, already provided by the New Brunswick parser);
3. "Maritime Link Import" for the Newfoundland interconnector.

The rest of the stuff in the issue is useful for exchange configuration and historical data.

Reply:
Awesome, thanks for clarifying! That doesn't sound as arduous as I thought; maybe I can roll it into this PR. I'll try to carve out some time to work on it this weekend.

Reply:
Alright, that wasn't so bad! A few notes:

- I created the `CA-NL-NF->CA-NS` exchange, but this is my first time doing so, so I very well could have done something wrong. In particular, I haven't tested how it will look because I don't have the front-end set up on my machine (I used Google Maps for the coordinates and a protractor for the rotation :grin:). Further, I wasn't sure whether to use `CA-NL-LB` or `CA-NL-NF` for the other zone (I chose the latter based on the physical location of the link).
- I also implemented the `CA-NB->CA-NS` exchange, but the back-end is still configured to use the NB parser for this pair.